### PR TITLE
#14474 Fix check component for privilege roles of postgresql

### PR DESCRIPTION
**Pull request by Lkledu**
This is proposed solution to #14474 
the issue was added to the 21.3.5 milestone

- Changed the `setPermission` method from `PostgrePrivilege` so it not overwrite the permissions anymore.
- Changed the `updateCurrentPrivileges` method from `PostgresRolePrivilegesEditor` so the `if(permission == null)` doesn't skip the addition of new permissions before user add the first permission.

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java [hunk 1]
```diff
@@ -163,9 +163,16 @@ public short getPermission(PostgrePrivilegeType privilegeType) {
...
+            if (permission.privilegeType != privilegeType) {
                 if (permit) {
                     permission.permissions |= GRANTED;
+                    ObjectPermission[] tempPermission = new ObjectPermission[this.permissions.length+1];
+                    for(int i = 0; i < this.permissions.length; i++) {
+                	tempPermission[i] = this.permissions[i];
+                    }
+                    tempPermission[this.permissions.length] = new ObjectPermission(privilegeType, permission.getGrantor(), permission.permissions);
+                    this.permissions = tempPermission;
+
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java]: Can be replaced with `org.jkiss.utils.ArrayUtils#add`.

File: plugins/org.jkiss.dbeaver.ext.postgresql.ui/src/org/jkiss/dbeaver/ext/postgresql/ui/editors/PostgresRolePrivilegesEditor.java [hunk 2]
```diff
@@ -413,7 +423,7 @@ private void updateObjectPermissions(List<DBSObject> objects) {
             for (TableItem item : permissionTable.getItems()) {
                 PostgrePrivilegeType privType = (PostgrePrivilegeType) item.getData();
                 short perm = currentPermissions[0] == null ? PostgrePrivilege.NONE : currentPermissions[0].getPermission(privType);
-                item.setChecked((perm & PostgrePrivilege.GRANTED) != 0);
+                item.setChecked((perm & PostgrePrivilege.GRANTED) != 0 );
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.postgresql.ui/src/org/jkiss/dbeaver/ext/postgresql/ui/editors/PostgresRolePrivilegesEditor.java]: Can be replaced with `org.jkiss.utils.CommonUtils#isBitSet`.

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java [hunk 3]
```diff
@@ -163,9 +163,16 @@ public short getPermission(PostgrePrivilegeType privilegeType) {
 
     public void setPermission(PostgrePrivilegeType privilegeType, boolean permit) {
         for (ObjectPermission permission : permissions) {
-            if (permission.privilegeType == privilegeType) {
+            if (permission.privilegeType != privilegeType) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java:180]: Is this change intended? This will bloat `permissions` with multiple `privilegeType`-s.

Also it updates `permissions` array while iterating through it which is bad.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java (hunk 1 above)
Comment (Lkledu) [reply plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java]: `org.jkiss.utils.ArrayUtils#add` just add int arrays, i've a ObjectPermission array

Comment (Lkledu) [reply plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java]: i change the for statement for `ArrayUtils.concatArrays()`

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java]: There's generic overload; usage: `ArrayUtils.add(ObjectPermission.class, permissions, new ObjectPermission(...))`.

Please see my other comment below.

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java (hunk 3 above)
Comment (Lkledu) [reply plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java:180]: it's intended. Before this implementation the permission array was overwritten, so alway when the userr change role and goes back, dbeaver would only check the last privilege checkbox, but let all the previous grant privilege statements on the permissionMap (so when user clicks on save button it shows all grant statement queries correct)

Comment (Lkledu) [reply plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgrePrivilege.java:180]: i had wrote a message before asking your opinion about using another object to store permissions list just for use with check UI instead to change the current permission object, but i don't know why the message get missed.
